tests-e2e/account_data_test.go

```diff
@@ -0,0 +1,160 @@
+package syncv3_test
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/sliding-sync/sync3"
+	"github.com/matrix-org/sliding-sync/sync3/extensions"
+	"github.com/matrix-org/sliding-sync/testutils"
+	"github.com/matrix-org/sliding-sync/testutils/m"
+	"testing"
+	"time"
+)
+
+func TestAccountDataRespectsExtensionScope(t *testing.T) {
+	alice := registerNewUser(t)
+
+	var syncResp *sync3.Response
+
+	// Want at least one test of the initial sync behaviour (which hits `ProcessInitial`)
+	// separate to the incremental sync behaviour (hits `AppendLive`)
+	t.Log("Alice creates rooms 1 and 2.")
+	room1 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 1"})
+	room2 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 2"})
+	t.Logf("room1=%s room2=%s", room1, room2)
+
+	t.Log("Alice uploads account data for both rooms, plus global account data.")
+	globalAccountDataEvent := putGlobalAccountData(
+		t,
+		alice,
+		"com.example.global",
+		map[string]interface{}{"global": "GLOBAL!", "version": 1},
+	)
+	putRoomAccountData(
+		t,
+		alice,
+		room1,
+		"com.example.room",
+		map[string]interface{}{"room": 1, "version": 1},
+	)
+	putRoomAccountData(
+		t,
+		alice,
+		room2,
+		"com.example.room",
+		map[string]interface{}{"room": 2, "version": 2},
+	)
+
+	t.Log("Alice makes an initial sync request, requesting global account data only.")
+	syncResp = alice.SlidingSync(t, sync3.Request{
+		Extensions: extensions.Request{
+			AccountData: &extensions.AccountDataRequest{
+				Core: extensions.Core{Enabled: &boolTrue, Lists: []string{}, Rooms: []string{}},
+			},
+		},
+		Lists: map[string]sync3.RequestList{
+			"window": {
+				Ranges: sync3.SliceRanges{{0, 20}},
+			},
+		},
+	})
+
+	t.Log("Alice should see her global account data only.")
+	m.MatchResponse(
+		t,
+		syncResp,
+		m.MatchHasGlobalAccountData(globalAccountDataEvent),
+		m.MatchNoRoomAccountData([]string{room1, room2}),
+	)
+
+	pos := syncResp.Pos
+	responses := make(chan *sync3.Response, 10)
+
+	waitForSyncResponse := func() *sync3.Response {
+		select {
+		case res := <-responses:
+			return res
+		case <-time.After(500 * time.Millisecond):
+			t.Fatalf("Timed out waiting for incremental sync response")
+		}
+		return nil
+	}
+
+	aliceIncrementalSync := func() {
+		t.Log("Alice incremental syncs, requesting account data for room 2")
+		response := alice.SlidingSync(
+			t,
+			sync3.Request{
+				Extensions: extensions.Request{
+					AccountData: &extensions.AccountDataRequest{
+						Core: extensions.Core{Enabled: &boolTrue, Lists: []string{}, Rooms: []string{room2}},
+					},
+				},
+			},
+			WithPos(pos),
+		)
+		pos = response.Pos
+		responses <- response
+	}
+
+	go aliceIncrementalSync()
```

Comment:
Please don't use goroutines unless you absolutely have to. You don't have to here. It's sufficient to just call `putGlobalAccountData` and then synchronously call `alice.SlidingSyncUntil(...)` to await for the response. This handles timeouts for you.

Reply:
The thinking here: I was trying to test that `AppendLive` has the correct logic in place. If the proxy is already aware of a change before the sync request comes in, that change gets picked up by `ProcessInitial` and a response is returned immediately, without hitting `AppendLive`.

Reply:
This thinking was not correct. See 391f0b8